### PR TITLE
Fixed issue #1755: Overload pcntl_fork() to prevent performance degradation by calling xdebug_get_pid often

### PR DESCRIPTION
**Pull request by carlos-granados**
In version 2.7 a change was introduced to restart the debugger if the process id changed. This was added to fix issue # 938 which stopped you from correctly debugging forked processes after using pnctl_fork() . The problem is that now we are calling xdebug_get_pid in a lot of places and this slows down the debugger a lot. It is not actually necessary to be calling xdebug_get_pid continually as we can just do that when the process is forked and restart the debugger at that point if needed. We do this by overriding the pnctl_fork() php function and checking the process id and restarting the debugger if needed there

By doing this, the measured time to run this php code when connected to a debugging client
```
<?php
function fib($n)
{
    if ($n <= 1) {
        return 1;
    }
    return fib($n - 1) + fib($n - 2);
}

echo fib(35);
```
decreased from 2:16 min to 0:59min

### CODE REVIEW

File: src/base/base.c [hunk 1]
```diff
@@ -513,6 +506,15 @@ static void xdebug_overloaded_functions_setup(void)
 	} else {
 		XG_BASE(orig_pcntl_exec_func) = NULL;
 	}
+
+	/* Override pcntl_fork with our own function to be able to start the debugger for the forked process */
```
Comment (derickr) [inline src/base/base.c]: Please line wrap at < 80 characters.

File: src/base/base.c [hunk 2]
```diff
@@ -635,7 +644,7 @@ void xdebug_base_rinit()
 	XG_BASE(filters_tracing)           = xdebug_llist_alloc(xdebug_llist_string_dtor);
 	XG_BASE(filters_code_coverage)     = xdebug_llist_alloc(xdebug_llist_string_dtor);
 
-	/* Overload var_dump, set_time_limit, error_reporting, and pcntl_exec */
+	/* Overload var_dump, set_time_limit, error_reporting, pcntl_exec and pcntl_fork */
```
Comment (derickr) [inline src/base/base.c]: Just remove this line, it doesn't say anything any more (After the big refactoring), and keeping it up to date is just annoying.

File: src/debugger/com.c [hunk 3]
```diff
@@ -453,13 +453,13 @@ int xdebug_is_debug_connection_active()
 	);
 }
 
-int xdebug_is_debug_connection_active_for_current_pid()
+void xdebug_activate_connection_for_current_pid()
```
Comment (derickr) [inline src/debugger/com.c]: The `base` code shouldn't really be calling this directly, and only code from within the `src/debugger` directory should. Instead, there should be an API function in `src/debugger/debugger.c` that calls this. Perhaps called `xdebug_debugger_restart_if_pid_changed`.

File: src/debugger/com.c [hunk 4]
```diff
@@ -453,13 +453,13 @@ int xdebug_is_debug_connection_active()
...
 {
 	zend_ulong pid;
 
 	/* Early return to save some getpid() calls */
```
Comment (derickr) [inline src/debugger/com.c]: This comment can go, as getpid is now only called once anyway.

File: xdebug.c [hunk 5]
```diff
@@ -807,15 +807,15 @@ PHP_FUNCTION(xdebug_set_time_limit)
...
+	if (ZEND_NUM_ARGS() == 0 && XG_BASE(error_reporting_overridden) && xdebug_is_debug_connection_active()) {
 		RETURN_LONG(XG_BASE(error_reporting_override));
 	}
 	XG_BASE(orig_error_reporting_func)(INTERNAL_FUNCTION_PARAM_PASSTHRU);
 }
 /* }}} */
 
 /* {{{ proto void xdebug_pcntl_exec(void)
-   Dummy function to prevent time limit from being set within the script */
+   Dummy function to stop profiling when we run pcntl_exec */
```
Comment (derickr) [inline xdebug.c]: Can you make this specific change in a separate commit? It's not related to this change. No need to create a separate pull request for it.

File: xdebug.c [hunk 6]
```diff
@@ -825,6 +825,16 @@ PHP_FUNCTION(xdebug_pcntl_exec)
 }
 /* }}} */
 
+/* {{{ proto int xdebug_pcntl_fork(void)
+   Dummy function to set a new connection when forking a process */
+PHP_FUNCTION(xdebug_pcntl_fork)
+{
+	XG_BASE(orig_pcntl_fork_func)(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+
+	xdebug_activate_connection_for_current_pid();
```
Comment (derickr) [inline xdebug.c]: This should instead call the new API function.

**Comment by carlos-granados** [comment]
Did all requested changes, thanks for your helpful comments. I am going to be away and unavailable until next Friday so I will not be able to answer any further questions or requests until then

**Comment by derickr** [comment]
I've merged this after a rebase and a few tweaks. Thanks! It's going to make it into Xdebug 3.0.
